nemo/collections/asr/models/msdd_models.py

```diff
@@ -0,0 +1,1736 @@
+                Shape: (batch_size, scale_n, number of base-scale segments, self.num_spks_per_model)
+            ms_seg_counts: (Tensor)
+                Cumulative sum of the number of segments in each scale. This information is needed to reconstruct
+                the multi-scale input matrix during forward propagating.
+                Shape: (batch_size, scale_n)
+
+        Returns:
+            ms_mel_feat: (Tensor)
+                Feature input stream split into the same length.
+                Shape: (total number of segments, feat_dim, self.frame_per_sec * the-longest-scale-length)
+            ms_mel_feat_len: (Tensor)
+                The actual length of feature without zero-padding.
+                Shape: (total number of segments,)
+            seq_len: (Tensor)
+                The length of the input embedding sequences.
+                Shape: (total number of segments,)
+            detach_ids: (tuple)
+                Tuple containing both detached embeding indices and attached embedding indices
+        """
+        device = torch.cuda.current_device()
+        _emb_batch_size = min(self.emb_batch_size, ms_seg_counts.sum().item() - self.min_detached_embs)
+        feat_dim = self.preprocessor._cfg.features
+        max_sample_count = int(self.multiscale_args_dict["scale_dict"][0][0] * self.frame_per_sec)
+        ms_mel_feat_len_list, sequence_lengths_list, ms_mel_feat_list = [], [], []
+        total_seg_count = torch.sum(ms_seg_counts)
+
+        batch_size = processed_signal.shape[0]
+        for batch_idx in range(batch_size):
+            for scale_idx in range(self.scale_n):
+                scale_seg_num = ms_seg_counts[batch_idx][scale_idx]
+                for k, (stt, end) in enumerate(ms_seg_timestamps[batch_idx][scale_idx][:scale_seg_num]):
+                    stt, end = int(stt.detach().item()), int(end.detach().item())
+                    end = min(end, stt + max_sample_count)
+                    _features = torch.zeros(feat_dim, max_sample_count).to(torch.float32).to(device)
+                    _features[:, : (end - stt)] = processed_signal[batch_idx][:, stt:end]
+                    ms_mel_feat_list.append(_features)
+                    ms_mel_feat_len_list.append(end - stt)
+            sequence_lengths_list.append(ms_seg_counts[batch_idx][-1])
+        ms_mel_feat = torch.stack(ms_mel_feat_list).to(device)
+        ms_mel_feat_len = torch.tensor(ms_mel_feat_len_list).to(device)
+        seq_len = torch.tensor(sequence_lengths_list).to(device)
+
+        torch.manual_seed(self._trainer.current_epoch)
+        if _emb_batch_size < self.min_detached_embs:
+            attached, _emb_batch_size = torch.tensor([]), 0
+            detached = torch.randperm(total_seg_count)
+        else:
+            attached = torch.randperm(total_seg_count)[:_emb_batch_size]
+            detached = torch.randperm(total_seg_count)[_emb_batch_size:]
+        detach_ids = (attached, detached)
```

Comment:
can you explain the logic to me,  of using detach_ids for eval and training of the speaker embedding model?

Reply:
This also need to be explained in person or a call